File: Confusing Number II/Confusing Number II.py
class Solution(object):
    def confusingNumberII(self, N):
        """
        :type N: int
        :rtype: int
        """
        d = {}
        d[0] = 0
        d[1] = 1
        d[6] = 9
        d[8] = 8
        d[9] = 6
        self.ans = 0
        
        
        def isCN(num):
            n,cn = num, 0
            while n>0:
                cn = cn*10+d[n%10]
                n //= 10
            return num!=cn
        
        def helper(N, cur):
            if isCN(cur):
                self.ans += 1
            
            for n in d:
                nextn = cur*10+d[n]
                if 1<=nextn<=N:
                    helper(N, nextn)
                    
        helper(N, 0)
        return self.ans

File: Confusing Number II/Confusing Number II.java
class Solution {
    HashMap<Integer, Integer> map = new HashMap();
    int res = 0;
    public int confusingNumberII(int N) {
        map.put(0,0);
        map.put(1,1);
        map.put(9,6);
        map.put(8,8);
        map.put(6,9);
        helper(N, 0);
        return res;
    }
    
    // 当前数字在个位加上所有confusing number，然后递归
    // 这道题没有回溯，只有递归
    private void helper(int N, long cur){
        if(isCN(cur)){
            res++;
        }
        
        for(int n: map.keySet()){
            long next = cur*10+n;
            if(next>=1 && next<=N){
                helper(N, next);
            }
        }
        
        return;
    }
    
    private boolean isCN(long num){
        long cn = 0;
        long n = num;
        while(n>0){
            cn = cn*10+ map.get((int)n%10);
            n /= 10;
        }
        return num!=cn;
    }
}